source: https://patchwork.ozlabs.org/patch/1045308/
From: Anton Gerasimov <tossel@gmail.com>
To: u-boot@lists.denx.de,
	agraf@suse.de
Subject: [U-Boot] [PATCH v2 1/1] board: raspberrypi: add serial and revision

Raspberry Pi bootloader adds this node to fdt, but if u-boot script
doesn't reuse the tree provided by it, this information is lost.

Revision and serial are displayed in /proc/cpuinfo after boot.

Suggested-By: Ricardo Salveti <rsalveti@rsalveti.net>
Reported-by: Karl Eaves <karleeaves@gmail.com>
Signed-off-by: Anton Gerasimov <tossel@gmail.com>
Reviewed-by: Alexander Graf <agraf@suse.de>

Index: git/board/raspberrypi/rpi/rpi.c
===================================================================
--- git.orig/board/raspberrypi/rpi/rpi.c
+++ git/board/raspberrypi/rpi/rpi.c
@@ -61,6 +61,8 @@ struct msg_get_clock_rate {
 	u32 end_tag;
 };
 
+static uint64_t serial;
+
 #ifdef CONFIG_ARM64
 #define DTB_DIR "broadcom/"
 #else
@@ -383,8 +385,9 @@ static void set_serial_number(void)
 		return;
 	}
 
-	snprintf(serial_string, sizeof(serial_string), "%016llx",
-		 msg->get_board_serial.body.resp.serial);
+	serial = msg->get_board_serial.body.resp.serial;
+	snprintf(serial_string, sizeof(serial_string), "%016llx", serial);
+
 	env_set("serial#", serial_string);
 }
 
@@ -481,6 +484,33 @@ void *board_fdt_blob_setup(void)
 }
 #endif
 
+static int ft_add_revision_info(void *blob) {
+	int off;
+	int ret;
+
+	off = fdt_subnode_offset(blob, 0, "system");
+
+	if (off < 0) {
+		off = fdt_add_subnode(blob, 0, "system");
+		if (off < 0)
+			return -1;
+	}
+
+	if (!fdt_getprop(blob, off, "linux,serial", NULL))  {
+		ret = fdt_setprop_u64(blob, off, "linux,serial", serial);
+		if (ret < 0)
+			return -1;
+	}
+
+	if (!fdt_getprop(blob, off, "linux,revision", NULL))  {
+		ret = fdt_setprop_u32(blob, off, "linux,revision", revision);
+		if (ret < 0)
+			return -1;
+	}
+
+	return 0;
+}
+
 int ft_board_setup(void *blob, bd_t *bd)
 {
 	/*
@@ -490,6 +520,8 @@ int ft_board_setup(void *blob, bd_t *bd)
 	 */
 	lcd_dt_simplefb_add_node(blob);
 
+	ft_add_revision_info(blob);
+
 #ifdef CONFIG_EFI_LOADER
 	/* Reserve the spin table */
 	efi_add_memory_map(0, 1, EFI_RESERVED_MEMORY_TYPE, 0);
